docs/state_encryption.md

```diff
@@ -0,0 +1,303 @@
+# State encryption
+
+This document details our intended implementation of the state and plan encryption feature.
+
+## Notes
+
+This document mentions `HCL` as a short-form for OpenTofu code. Unless otherwise specified, everything written also applies to the [JSON-equivalent of the HCL code](https://opentofu.org/docs/language/syntax/json/).
+
+## Goals
+
+The goal of this feature is to allow OpenTofu users to fully encrypt state files when they are stored on the local disk or transferred to a remote backend. The feature should also allow reading from an encrypted remote backend using the `terraform_remote_state` data source. The encrypted version should still be a valid JSON file, but not necessarily a valid state file.
+
+Furthermore, this feature should allow users to encrypt plan files when they are stored. However, plan files are not JSON, so there is no expectation of the resulting plan file being JSON either. **TODO: do we want to specify that we use ZIP encryption semantics?**
+
+For the encryption key, users should be able to specify a key directly, use a remote key provider (such as AWS KMS, etc.), or create derivative keys (such as pbkdf2) from another key source. The primary encryption method should be AES-GCM, but the implementation should be open to different encryption methods. The user should also have the ability to decrypt a state file with one (older) key and then re-encrypt data with a newer key.
+
+To enable use cases where multiple teams need to collaborate, the user should be able to specify separate encryption methods and keys for individual uses, especially for the `terraform_remote_state` data source. However, to simplify configuration, the user should be able to specify single configuration for all remote state data sources.
+
+It is the goal of this feature to let users specify their encryption configuration both in HCL and in environment variables. The latter is necessary to allow users the reuse of HCL code for encrypted and unencrypted state storage.
+
+Finally, it is the goal of the encryption feature to make available a library that third party tooling can use to encrypt and decrypt state. This may be implemented as a package within the opentofu repository, or as a standalone repository.
+
+## Future goals
+
+In the future, the user should be able to specify partial encryption. This encryption type would only encrypt sensitive values instead of the whole state file. **TODO how about the plan file?**
```

Comment:
I think partial encryption should only apply to state files. Plans are binary data.